### PR TITLE
1010 resolution scaling

### PR DESCRIPTION
**Pull request by bodamat**
Resolution scale and post-process change when fps slow or backwards. Check to please that works ok

I see flashing at the sharp change of post-process. I'm not sure it's ok

### CODE REVIEW

File: packages/engine/src/renderer/systems/WebGLRendererSystem.ts [hunk 1]
```diff
@@ -159,16 +182,76 @@ export class WebGLRendererSystem extends System {
...
+  changeQualityLevel(delta: number) {
+    // start timer when fps changes
+    if (this.fps <= 45) {
+      this.downGradeTimer += delta;
+      this.upGradeTimer = 0;
+    }
+    else if (this.fps >= 55) {
+      this.upGradeTimer += delta;
+      this.downGradeTimer = 0;
+    }
```
Comment (XiaNi) [inline packages/engine/src/renderer/systems/WebGLRendererSystem.ts:224]: reset both timers if fps is not lower 45 and not higher then 55
```
    else {
      this.upGradeTimer = 0;
      this.downGradeTimer = 0;
    }
```

File: packages/engine/src/renderer/systems/WebGLRendererSystem.ts [hunk 2]
```diff
@@ -159,16 +182,76 @@ export class WebGLRendererSystem extends System {
...
         resize(entity);
-        getComponent<RendererComponent>(entity, RendererComponent).composer.render(delta);
+
+        if (this.qualityLevel >= 2) {
+          getComponent<RendererComponent>(entity, RendererComponent).composer.render(delta);
+          Engine.renderer.outputEncoding = LinearEncoding; // need this if postprocessing is used
+        }
+        else {
+          Engine.renderer?.render(Engine.scene, Engine.camera);
+          Engine.renderer.outputEncoding = sRGBEncoding; // need this if postprocessing is not used
```
Comment (XiaNi) [inline packages/engine/src/renderer/systems/WebGLRendererSystem.ts:206]: if in line 205 you are not sure that Engine.renderer is set, then add same optional chaining in line 206.
Engine.renderer?.outputEncoding = sRGBEncoding;

Comment (bodamat) [reply packages/engine/src/renderer/systems/WebGLRendererSystem.ts:206]: unfortunately, it throws an error in line 206

Comment (bodamat) [reply packages/engine/src/renderer/systems/WebGLRendererSystem.ts:206]: The left-hand side of an assignment expression may not be an optional property access.

Comment (XiaNi) [reply packages/engine/src/renderer/systems/WebGLRendererSystem.ts:206]: oh, sure, than
```js
if (Engine.renderer) {
          Engine.renderer.render(Engine.scene, Engine.camera);
          Engine.renderer.outputEncoding = sRGBEncoding; 
}
```

Comment (bodamat) [reply packages/engine/src/renderer/systems/WebGLRendererSystem.ts:206]: Thanks. This is better solution to the problem

File: packages/engine/src/renderer/systems/WebGLRendererSystem.ts [hunk 3]
```diff
@@ -159,16 +183,89 @@ export class WebGLRendererSystem extends System {
...
+        case 3:
+          this.scaleFactor = 1;
+          break;
+        default:
+          this.scaleFactor = 1;
+          break;
+      }
+
+      Engine.renderer?.setPixelRatio(window.devicePixelRatio * this.scaleFactor);
+      this.prevQualityLevel = this.qualityLevel;
```
Comment (XiaNi) [inline packages/engine/src/renderer/systems/WebGLRendererSystem.ts:265]: change it to
```js
if (Engine.renderer) {
      Engine.renderer.setPixelRatio(window.devicePixelRatio * this.scaleFactor);
      this.prevQualityLevel = this.qualityLevel;
}
```

**Review by XiaNi — APPROVED** [review]

**Pull request merged**